janitor/io.py

```diff
@@ -69,3 +71,45 @@ def read_csvs(
         )
     else:
         return dfs_dict
+
+
+def read_commandline(cmd: str) -> pd.DataFrame:
+    """
+    Read a CSV file based on a command-line command.
+
+    For example, you may wish to run the following command on `sep-quarter.csv`
+    before reading it into a pandas DataFrame:
+
+    ```bash
+    cat sep-quarter.csv | grep .SEA1AA
+    ```
+
+    In this case, you can use the following Python code to load the dataframe:
+
+    ```python
+    import janitor as jn
+    df = jn.read_commandline("cat sep-quarter.csv | grep .SEA1AA")
+    ```
+
+    :param cmd: Shell command to preprocess a file on disk.
+    :returns: A pandas DataFrame parsed from the stdout of the underlying
+        shell.
+    :raises EmptyDataError: If there is no data to parse, this often happens
+        because the cmd param is either an invalid bash command, thus
+        nothing happens in the shell , or if cmd param is not a string,
+        thus creating an invalid shell command.
+    """
+    # cmd = cmd.split(" ")
+    try:
+        outcome = subprocess.run(
+            cmd, shell=True, capture_output=True, text=True
+        )
+        outcome = outcome.stdout
+    except pd.EmptyDataError:
```

Comment:
Could you explain when such an exception will occur? I doubt it will, right? Since there is nothing related to `pandas` in the try block. I'm guessing this is leftover from previous code. Or possibly you meant to have the `pd.read_csv` inside the try block. (btw, there is no `pd.EmptyDataError`, it's `pd.errors.EmptyDataError`.)

Furthermore, I'm thinking whether it is a better user experience if we can also propagate the stderr back to the user IF return_code is non-zero, so the user explicitly knows that the error is because of the shell command being malformed etc., instead of a pandas read error.

---

Also, I'm not convinced that we should even be catching empty data error; that is functionality that is passed to pandas, we can simply let pandas throw the error in a transparent manner, IMO.

For example:
```python
>>> janitor.io.read_commandline("")
```
<img width="1010" alt="image" src="https://user-images.githubusercontent.com/30731072/148780377-a3cc2580-96c6-4f8b-b48d-f97fd1f3b473.png">

I think this stacktrace is clear enough, we shouldn't need to catch and re-raise the exception?

Reply:
That was my thought too, I think Pandas throws an error before it even finishes parsing through the try block, which would indeed do nothing more than re-raising an exception that was already raised if that code is left in.